### PR TITLE
feat: Add more unit tests

### PR DESCRIPTION
**Pull request by toVersus**
Sorry for bothering you. I completely understand how to collaborate with others on the same PR :)

## Proposed Changes

* Fix neat comments and others
* Output unit test name for each run
* Add more unit tests

### CODE REVIEW

File: pkg/kn/traffic/compute.go [hunk 1]
```diff
@@ -85,8 +85,7 @@ func (e ServiceTraffic) IsTagPresent(tag string) bool {
 func (e ServiceTraffic) UntagRevision(tag string) {
 	for i, target := range e {
 		if target.Tag == tag {
-			target.Tag = ""
-			e[i] = target
+			e[i].Tag = ""
```
Comment (navidshaikh) [inline pkg/kn/traffic/compute.go:88]: :+1:

File: pkg/kn/traffic/compute.go [hunk 2]
```diff
@@ -303,7 +295,7 @@ func Compute(cmd *cobra.Command, targets []v1alpha1.TrafficTarget, trafficFlags
 			revisionRef, percent, _ := splitByEqualSign(each) // err is verified in verifyInputSanity
 			percentInt, _ := strconv.Atoi(percent)            // percentInt (for int) is verified in verifyInputSanity
 
-			// fourth precendence: set traffic for latest revision
+			// fourth precedence: set traffic for latest revision
```
Comment (navidshaikh) [inline pkg/kn/traffic/compute.go:298]: thanks!

File: pkg/kn/traffic/compute_test.go [hunk 3]
```diff
@@ -25,6 +25,7 @@ import (
 )
 
 type trafficTestCase struct {
+	name             string
```
Comment (navidshaikh) [inline pkg/kn/traffic/compute_test.go:28]: that's something I wanted to bring in as subsequent change, thats for thinking through this and improving it.

File: pkg/kn/traffic/compute_test.go [hunk 4]
```diff
@@ -51,169 +53,188 @@ func newTestTrafficCommand() (*cobra.Command, *flags.Traffic) {
...
+			"untag 'latest' tag from 'echo-v1' revision",
 			append(newServiceTraffic([]v1alpha1.TrafficTarget{}), newTarget("latest", "echo-v1", 100, false)),
 			[]string{"--untag", "latest"},
 			[]string{"echo-v1"},
 			[]string{""},
 			[]int{100},
 		},
+		{
+			"untag 'latest' tag from '@latst' revision",
+			append(newServiceTraffic([]v1alpha1.TrafficTarget{}), newTarget("latest", "", 100, true)),
```
Comment (navidshaikh) [inline pkg/kn/traffic/compute_test.go]: this test doesn't check anything additional TBH as untagging is unconditional whether its latest or non-latest tag.

I tried to keep the tests compact to avoid more tests, the coverage is already 100%

as we don't have latest revision check in Untag method, I'd suggest we more this test as its just duplicate of earlier one.

File: pkg/kn/traffic/compute_test.go [hunk 5]
```diff
@@ -51,169 +53,188 @@ func newTestTrafficCommand() (*cobra.Command, *flags.Traffic) {
...
+			[]string{"--untag", "latest"},
+			[]string{"@latest"},
+			[]string{""},
+			[]int{100},
+		},
+		{
+			"replace revision pointing to 'latest' tag from 'echo-v1' to 'echo-v2' revision",
+			append(newServiceTraffic([]v1alpha1.TrafficTarget{}), newTarget("latest", "echo-v1", 50, false), newTarget("", "echo-v2", 50, false)),
+			[]string{"--untag", "latest", "--tag", "echo-v1=old,echo-v2=latest"},
```
Comment (navidshaikh) [inline pkg/kn/traffic/compute_test.go:147]: again, doesn't additionally test any additional combination, the individual operations are tested in earlier tests with different combinations.

File: pkg/kn/traffic/compute_test.go [hunk 6]
```diff
@@ -51,169 +53,188 @@ func newTestTrafficCommand() (*cobra.Command, *flags.Traffic) {
...
-			t.Fatal(err)
-		}
-		for i, target := range targets {
-			if testCase.desiredRevisions[i] == "@latest" {
-				assert.Equal(t, *target.LatestRevision, true)
-			} else {
-				assert.Equal(t, target.RevisionName, testCase.desiredRevisions[i])
+		t.Run(testCase.name, func(t *testing.T) {
+			if lper, lrev, ltag := len(testCase.desiredPercents), len(testCase.desiredRevisions), len(testCase.desiredTags); lper != lrev || lper != ltag {
+				t.Fatalf("length of desird revisions, tags and percents is mismatched: got=(desiredPercents, desiredRevisions, desiredTags)=(%d, %d, %d)",
```
Comment (navidshaikh) [inline pkg/kn/traffic/compute_test.go:155]: *desired

Comment (navidshaikh) [reply pkg/kn/traffic/compute_test.go:155]: We don't need this, this would raise dev-time list index out of range error which dev should fix.

And as u have added the test case names, that should be easier to check, I'd recommend removing the len check altogether.

WDYT ?

File: pkg/kn/traffic/compute_test.go [hunk 7]
```diff
@@ -51,169 +53,188 @@ func newTestTrafficCommand() (*cobra.Command, *flags.Traffic) {
...
-		// error for --traffic for invalid split operation
 		{
+			"invalid format for `--traffic` option",
 			append(newServiceTraffic([]v1alpha1.TrafficTarget{}), newTarget("", "", 100, true)),
 			[]string{"--traffic", "@latest=100=latest"},
 			"expecting the value format in value1=value2, given @latest=100=latest",
 		},
-		// error for --tag for invalid split operation
 		{
+			"invalid format for `--tag` option",
```
Comment (navidshaikh) [inline pkg/kn/traffic/compute_test.go]: we used ` here for --tag

File: pkg/kn/traffic/compute_test.go [hunk 8]
```diff
@@ -51,169 +53,188 @@ func newTestTrafficCommand() (*cobra.Command, *flags.Traffic) {
...
 			append(newServiceTraffic([]v1alpha1.TrafficTarget{}), newTarget("", "", 100, true)),
 			[]string{"--tag", "@latest="},
 			"expecting the value format in value1=value2, given @latest=",
 		},
-		// error for --traffic for repeating @latest
 		{
+			"repeatedly spliting traffic to '@latest' revision",
```
Comment (navidshaikh) [inline pkg/kn/traffic/compute_test.go]: and '' for @latest here.

I'd say lets remove both of them and use them as plain strings like --tag and @latest

File: pkg/kn/traffic/compute_test.go [hunk 9]
```diff
@@ -51,169 +53,188 @@ func newTestTrafficCommand() (*cobra.Command, *flags.Traffic) {
...
-		// error for overwriting tag for @latest
 		{
+			"overwriting tag to '@latest' revision not allowed",
 			append(append(newServiceTraffic([]v1alpha1.TrafficTarget{}), newTarget("latest", "", 2, true)), newTarget("stable", "echo-v2", 98, false)),
 			[]string{"--tag", "@latest=stable"},
 			"refusing to overwrite existing tag in service, add flag '--untag stable' in command to untag it",
 		},
-		// error for overwriting tag for revision
 		{
+			"overwriting tag to same revision not allowed",
```
Comment (navidshaikh) [inline pkg/kn/traffic/compute_test.go]: *overwriting tags of others revisions not allowed

File: pkg/kn/traffic/compute_test.go [hunk 10]
```diff
@@ -51,169 +53,188 @@ func newTestTrafficCommand() (*cobra.Command, *flags.Traffic) {
...
 			append(append(newServiceTraffic([]v1alpha1.TrafficTarget{}), newTarget("latest", "", 2, true)), newTarget("stable", "echo-v2", 98, false)),
 			[]string{"--tag", "echo-v2=latest"},
 			"refusing to overwrite existing tag in service, add flag '--untag latest' in command to untag it",
 		},
-		// error for Atoi
 		{
+			"invalid format for traffic percentage",
```
Comment (navidshaikh) [inline pkg/kn/traffic/compute_test.go]: verify error for non integer values given to percent

**Review by navidshaikh — CHANGES_REQUESTED** [review]
Thanks, looks good to me with few changes requested.
Please squash in next update.

File: pkg/kn/traffic/compute_test.go (hunk 3 above)
Comment (toVersus) [reply pkg/kn/traffic/compute_test.go:28]: Yeah, I just introduced a simple solution for identifying a failed test scenario so that we can improve this further as we like.

File: pkg/kn/traffic/compute_test.go (hunk 4 above)
Comment (toVersus) [reply pkg/kn/traffic/compute_test.go]: > as we don't have latest revision check in Untag method,

Ah, that makes sense.

File: pkg/kn/traffic/compute_test.go (hunk 5 above)
Comment (toVersus) [reply pkg/kn/traffic/compute_test.go:147]: > the individual operations are tested in earlier tests with different combinations.

That's true, and my motivation to introduce this test case is that I would like to cover all the combinations described in Example section in the [docs](https://github.com/knative/client/pull/331).

File: pkg/kn/traffic/compute_test.go (hunk 6 above)
Comment (toVersus) [reply pkg/kn/traffic/compute_test.go:155]: Yeah, it may easy to find due to the test case names, but IMO it scared me and made me suspect that we have a potential bug in our core implementations (not test code). This  simple length check shows me the route cause of this index-out-of-range error at a glance.

File: pkg/kn/traffic/compute_test.go (hunk 5 above)
Comment (navidshaikh) [reply pkg/kn/traffic/compute_test.go:147]: Fine.

I planned to cover the examples in docs in e2e and have unit tests cover to *unit* functionality. I am fine with keeping it.

File: pkg/kn/traffic/compute_test.go (hunk 6 above)
Comment (navidshaikh) [reply pkg/kn/traffic/compute_test.go:155]: okay, TBH I should've added code comments in tests about this requirement.

**Review by navidshaikh — APPROVED** [review]
ready to merge, can you please squash and update?

File: pkg/kn/traffic/compute_test.go (hunk 3 above)
Comment (navidshaikh) [reply pkg/kn/traffic/compute_test.go:28]: Percent. :+1:

**Comment by navidshaikh** [comment]
oh, github gives me option to squash and merge as well. nvm!
